### ISSUE TITLE
@PathParam being wrongly added to other methods

### ISSUE THREAD

**Issue by waggy55** [smallrye/smallrye-open-api#199]
Using Quarkus 0.26.1, and given the following 2 end-points, one of which has the @PathParam, `productId`:

```
//Class defined: @Path("/products")

@GET()
@Path("")
public CompletionStage<Response> getAllProducts()
{  /* SOME CODE */  }

and

@GET()
@Path("/{productId}")
public CompletionStage<Response> getProduct(@PathParam("productId") String productId)
{  /* SOME MORE CODE */  }
```

I get the following open api - as expected because only `/products/{productId}` has the parameter `productId`
```
openapi: 3.0.1
info:
  title: Generated API
  version: "1.0"
paths:
  /products:
    get:
      responses:
        200:
          description: OK
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Response'
  /products/{productId}:
    get:
      parameters:
      - name: productId
        in: path
        required: true
        schema:
          type: string
      responses:
        200:
          description: OK
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Response'
```

However, if I remove the second method and add the following method (if I leave the second method there and add the following method, the result is the same):
```
@POST()
@Path("/{productId}/reserve/")
public CompletionStage<Response> reserveProduct(@PathParam("productId") String productId, int quantity)
{  /* SOME CODE */  }
```

The @PathParam, `productId`, is **wrongly** applied to the first end-point, `/products`:
```
openapi: 3.0.1
info:
  title: Generated API
  version: "1.0"
paths:
  /products:
    get:
      responses:
        200:
          description: OK
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Response'
    parameters:
    - name: productId
      in: path
      required: true
      schema:
        type: string
  /products/{productId}/reserve:
    post:
      parameters:
      - name: productId
        in: path
        required: true
        schema:
          type: string
      requestBody:
        content:
          application/json:
            schema:
              format: int32
              type: integer
      responses:
        200:
          description: OK
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Response'
    parameters:
    - name: productId
      in: path
      required: true
      schema:
        type: string
```

In both instances, the `/products` end-point does not have a @PathParam, but in the second instance, the @PathParam from `/products/{productId}/reserve` is applied to `/products` - presumably this is due to the order in which the class's methods are evaluated.

Many thanks,
Dan

**Comment by MikeEdgar** [comment]
@waggy55 , when you removed the second method, did you fully remove the code or only comment out the JAX-RS annotations like `@GET` and `@Path` ?

**Comment by waggy55** [comment]
Hi @MikeEdgar, I removed the code fully (method and annotations)

**Comment by MikeEdgar** [comment]
With the following code, I am able to reproduce your issue - hence my earlier question. However, if the entire middle method is removed, I am not able to reproduce it. Are there any other methods or fields in your resource?

```java
@Path("/products")
static class SmallRye199Resource {
    @GET()
    @Path("")
    public CompletionStage<Response> getAllProducts()
    {  return null;  }

    //@GET()
    //@Path("/{productId}")
    public CompletionStage<Response> getProduct(@PathParam("productId") String productId)
    {  return null;  }

    @POST()
    @Path("/{productId}/reserve/")
    public CompletionStage<Response> reserveProduct(@PathParam("productId") String productId, int quantity)
    {  return null;  }
}
```

**Comment by waggy55** [comment]
There were, but I commented them out.

Here's the full class:

``` java
@Path("/products")
@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
public class ProductController
{
  private static final Logger LOGGER = LoggerFactory.getLogger(ProductController.class);
  private final ProductService productService;

  @Inject
  public ProductController(ProductService productService)
  {
    this.productService = productService;
  }

  @GET()
  @Path("")
  public CompletionStage<Response> getAllProducts()
  {
    return null;
  }

  @GET()
  @Path("/{productId}")
  public CompletionStage<Response> getProduct(@PathParam("productId") String productId)
  {
    return null;
  }

  @POST()
  @Path("")
  public CompletionStage<Response> addProduct(ProductIdDto productIdDto)
  {
    return null;
  }

  @POST()
  @Path("/{productId}/reserve/")
  public CompletionStage<Response> reserveProduct(@PathParam("productId") String productId, int quantity)
  {
    return null;
  }

  @GET()
  @Path("/reservations/{productId}")
  public CompletionStage<Response> getAllReservations(@PathParam("productId") String productId)
  {
    return null;
  }
}
```

Interestingly, when I replace all methods' code with `return null`, the @PathParams are as expected, i.e. `/products` does not require `productId`

However, if I add dummy code to any method (`getProduct(...)`, `reserveProduct(...)`, `getAllReservations(...)`) that uses `@PathParam("productId") String productId`, `/products` then wrongly expects the param `productId`

```java
return CompletableFuture.supplyAsync(() -> {
  LOGGER.info("POST Reserve Product {}", UUID.fromString(productId));
  return Response.serverError().build();
});
```

**Comment by MikeEdgar** [comment]
@waggy55 - I have not been able to reproduce this. If you have some time to put a "broken" version in a Github repository, that would be helpful.

**Comment by waggy55** [comment]
@MikeEdgar - I've uploaded a slim-downed version of the project: https://github.com/waggy55/quarkus-broken-pathparams

I'm running with the command `mvn compile quarkus:dev`

Sure enough, PathParam `productId` is applied to both end points
```
openapi: 3.0.1
info:
  title: Generated API
  version: "1.0"
paths:
  /products:
    get:
      responses:
        200:
          description: OK
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Response'
    parameters:
    - name: productId
      in: path
      required: true
      schema:
        type: string
  /products/{productId}:
    get:
      parameters:
      - name: productId
        in: path
        required: true
        schema:
          type: string
      responses:
        200:
          description: OK
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Response'
    parameters:
    - name: productId
      in: path
      required: true
      schema:
        type: string
```

Many thanks,
Dan

**Comment by MikeEdgar** [comment]
@waggy55 , believe it or not, it's still not happening on my machine with that project. Would you mind removing any locally downloaded smallrye-open-api artifacts from your Maven repository and retrying the `mvn` command? That is, anything here: `~/.m2/repository/io/smallrye/smallrye-open-api*`. Also, please reply with which JDK version you're using, in case that ends up making some difference.

Thanks

**Comment by waggy55** [comment]
So, GIVEN JDK 1.8.0_111
AND I've deleted my project folder and everything in `/.m2/repository/io/smallrye/`,
WHEN I cloned https://github.com/waggy55/quarkus-broken-pathparams,
AND then I ran the project using, `mvn clean compile quarkus:dev`
THEN the problem remained

However, if I use JDK jdk1.8.0_231, the problem goes away!

Apologies if I've missed something that said I shouldn't be using 1.8.0_111

Thank you very much indeed for your time!!

**Comment by MikeEdgar** [comment]
This is definitely unexpected behavior (for me). I confirmed also with that version of the JDK (1.8.0_111). Note that it did not occur by simply running the application, I had to compile the application for it to happen (`mvn clean`). Compiling with 1.8.0_111 and running with a new JRE also causes the problem. This sounds like the older JDK had a bug with how it was generating bytecode, which is what SmallRye ultimately is using to do the scan.

Thanks for the report. I don't think we need to do anything further for this issue, other than be aware that it could occur.
